explainability-service/src/main/java/org/kie/trustyai/service/endpoints/metrics/drift/KSTestEndpoint.java

```diff
@@ -0,0 +1,96 @@
+    }
+
+    private static final Logger LOG = Logger.getLogger(KSTestEndpoint.class);
+
+    // === THRESHOLDS ======================================
+    // determine if the metric value(s) exceed the provided threshold(s)
+    @Override
+    public MetricThreshold thresholdFunction(Number delta, MetricValueCarrier metricValue) {
+        return new MetricThreshold(
+                0,
+                delta.doubleValue(),
+                metricValue.getNamedValues().values().stream().max(Comparator.naturalOrder()).orElse(0.));
+    }
+
+    @Override
+    public String getGeneralDefinition() {
+        return "KSTest calculates two sample kolmogorov-smirnov test per column which tests if two samples drawn from the same distributions.";
+    }
+
+    // a specific definition for this value of this metric in this specific context
+    @Override
+    public String getSpecificDefinition(MetricValueCarrier metricValues, @ValidDriftMetricRequest KSTestMetricRequest request) {
+        StringBuilder out = new StringBuilder(getGeneralDefinition());
+        out.append(System.getProperty("line.separator"));
+
+        int maxColLen = metricValues.getNamedValues().keySet().stream().mapToInt(s -> s.length()).max().orElse(0);
+        String fmt = String.format("%%%ds", maxColLen);
+
+        for (Map.Entry<String, Double> entry : metricValues.getNamedValues().entrySet()) {
+            boolean reject = entry.getValue() <= request.getThresholdDelta();
+            out.append(String.format("  - Column %s has p=%f probability of coming from the training distribution.",
+                    String.format(fmt, entry.getKey()),
+                    entry.getValue()));
+            if (reject) {
+                out.append(String.format(" p <= %f -> [SIGNIFICANT DRIFT]", request.getThresholdDelta()));
+            } else {
+                out.append(String.format(" p >  %f", request.getThresholdDelta()));
+            }
+            out.append(System.getProperty("line.separator"));
+        }
+        return out.toString();
+    }
+
+    @Override
+    @CacheResult(cacheName = "metrics-calculator-kstest", keyGenerator = MetricCalculationCacheKeyGen.class)
+    public MetricValueCarrier calculate(Dataframe dataframe, @ValidReconciledMetricRequest BaseMetricRequest request) {
+        KSTestMetricRequest ksRequest = (KSTestMetricRequest)request;
+        KSTest ks = new KSTest();
+        // Get train data
+        Dataframe trainData = super.dataSource.get()
+                    .getDataframe(request.getModelId())
+                    .filterRowsByTagEquals(ksRequest.getReferenceTag());
+        // get data that does _not_ have the provided reference tag: test data
+        Dataframe filtered = dataframe.filterRowsByTagNotEquals(ksRequest.getReferenceTag());
+        LOG.debug("Cache miss. Calculating metric for " + ksRequest.getModelId());
+        Map<String, HypothesisTestResult> result = ks.calculate(trainData, filtered, ksRequest.getThresholdDelta());
```

Comment:
Probably same thing here if trainData or filtered sizes are less than 2, we need a safe execution path